services/cd-service/pkg/sqlitestore/sqlite.c

```diff
@@ -0,0 +1,212 @@
+	return found;
+}
+
+
+int kp_backend__write(git_odb_backend *_backend, const git_oid *id, const void *data, size_t len, git_otype type)
+{
+	int error;
+	kp_backend *backend;
+
+
+	backend = (kp_backend *)_backend;
+
+	error = SQLITE_ERROR;
+
+	if (sqlite3_bind_text(backend->write, 1, (char *)id->id, 20, SQLITE_TRANSIENT) == SQLITE_OK &&
+		sqlite3_bind_int(backend->write, 2, (int)type) == SQLITE_OK &&
+		sqlite3_bind_int(backend->write, 3, len) == SQLITE_OK &&
+		sqlite3_bind_blob(backend->write, 4, data, len, SQLITE_TRANSIENT) == SQLITE_OK) {
+		error = sqlite3_step(backend->write);
+	}
+
+	sqlite3_reset(backend->write);
+	return (error == SQLITE_DONE) ? GIT_OK : GIT_ERROR;
+}
+
+
+void kp_backend__free(git_odb_backend *_backend)
+{
+	kp_backend *backend;
+	backend = (kp_backend *)_backend;
+
+	sqlite3_finalize(backend->read);
+	sqlite3_finalize(backend->read_header);
+	sqlite3_finalize(backend->write);
+	sqlite3_close(backend->db);
+
+	free(backend);
+}
+
+static int create_table_if_not_exists(sqlite3 *db)
+{
+	return sqlite3_exec(db, 
+                "CREATE TABLE IF NOT EXISTS 'odb' ("
+		"'oid' CHARACTER(20) PRIMARY KEY NOT NULL,"
+		"'type' INTEGER NOT NULL,"
+		"'size' INTEGER NOT NULL,"
+		"'data' BLOB);", NULL, NULL, NULL);
+}
+
+static int init_statements(kp_backend *backend)
+{
+        int error = SQLITE_ERROR;
+	error = sqlite3_prepare_v2(backend->db,
+			    "SELECT type, size, data FROM 'odb' WHERE oid = ?;"
+			    , -1, &backend->read, NULL);
+        if( error != SQLITE_OK ){
+		return error;
+        }
+        error = sqlite3_prepare_v2(backend->db,
+				   "SELECT type, size FROM 'odb' WHERE oid = ?;",
+				   -1, &backend->read_header, NULL);
+        if( error != SQLITE_OK ){
+		return error;
+        }
+	return sqlite3_prepare_v2(backend->db,
+			   "INSERT OR IGNORE INTO 'odb' VALUES (?, ?, ?, ?);",
+			   -1, &backend->write, NULL);
+}
+
+int kp_backend_sqlite(git_odb_backend **backend_out, const char *sqlite_db, const char **err_out)
+{
+	int error = SQLITE_ERROR;
+	kp_backend *backend = calloc(1, sizeof(kp_backend));
+	if (backend == NULL) {
+		giterr_set_oom();
+		return SQLITE_ERROR;
+	}
+
+        error = sqlite3_open(sqlite_db, &backend->db);
+	if (error != SQLITE_OK)
+		goto cleanup;
+
+	error = create_table_if_not_exists(backend->db);
+	if (error != SQLITE_OK)
+		goto cleanup;
+
+	error = init_statements(backend);
+	if (error != SQLITE_OK)
+		goto cleanup;
+
+	backend->parent.version = GIT_ODB_BACKEND_VERSION;
+	backend->parent.read = &kp_backend__read;
+	backend->parent.read_prefix = &kp_backend__read_prefix;
+	backend->parent.read_header = &kp_backend__read_header;
+	backend->parent.write = &kp_backend__write;
+	backend->parent.exists = &kp_backend__exists;
+	backend->parent.free = &kp_backend__free;
+
+	*backend_out = (git_odb_backend *)backend;
+        return SQLITE_OK;
+cleanup:
+        *err_out = sqlite3_errmsg(backend->db);
+	kp_backend__free((git_odb_backend *)backend);
+	return error;
+}
```

Comment:
Overall, I think this is okay for now.
However, I am still a bit hesitant to merge C code. I don't see this as a long term solution.
-> We should think about how this should work long term.

Reply:
discussed for now: we need to enable devs to debug the C parts